pydantic/_internal/_std_types_schema.py

```diff
@@ -307,141 +294,296 @@ def validate(self, __input_value: int | float | str) -> Decimal:  # noqa: C901 (
-    inner_schema: core_schema.CoreSchema | None = None,
-) -> core_schema.WrapSerializerFunctionSerSchema:
-    return core_schema.wrap_serializer_function_ser_schema(
-        _serializers.serialize_deque, info_arg=True, schema=inner_schema or core_schema.any_schema()
-    )
 
+@slots_dataclass
+class SequenceValidator:
+    mapped_origin: type[Any]
+    item_source_type: type[Any]
+    min_length: int | None = None
+    max_length: int | None = None
+    strict: bool = False
+    js_core_schema: CoreSchema | None = None
+
+    def __get_pydantic_json_schema__(self, _schema: CoreSchema, handler: GetJsonSchemaHandler) -> JsonSchemaValue:
+        assert self.js_core_schema is not None
+        return handler(self.js_core_schema)
+
+    def serialize_sequence_via_list(
+        self, v: Any, handler: core_schema.SerializerFunctionWrapHandler, info: core_schema.SerializationInfo
+    ) -> Any:
+        items: list[Any] = []
+        for index, item in enumerate(v):
+            try:
+                v = handler(item, index)
+            except PydanticOmit:
+                pass
+            else:
+                items.append(v)
+
+        if info.mode_is_json():
+            return items
+        else:
+            return self.mapped_origin(items)
 
-def _deque_any_schema() -> core_schema.LaxOrStrictSchema:
-    metadata = build_metadata_dict(js_functions=[lambda _c, h: h(core_schema.list_schema(core_schema.any_schema()))])
-    lax_schema = core_schema.no_info_wrap_validator_function(
-        _validators.deque_any_validator,
-        core_schema.list_schema(),
-    )
-    strict_schema = core_schema.json_or_python_schema(
-        json_schema=lax_schema,
-        python_schema=core_schema.is_instance_schema(deque),
-    )
-    return core_schema.lax_or_strict_schema(
-        lax_schema=lax_schema,
-        strict_schema=strict_schema,
-        serialization=_deque_ser_schema(),
-        metadata=metadata,
-    )
+    def __get_pydantic_core_schema__(self, source_type: Any, handler: GetCoreSchemaHandler) -> CoreSchema:
+        if self.item_source_type is Any:
+            items_schema = None
+        else:
+            items_schema = handler.generate_schema(self.item_source_type)
 
+        metadata = {'min_length': self.min_length, 'max_length': self.max_length, 'strict': self.strict}
 
-@schema_function(deque)
-def deque_schema(schema_generator: GenerateSchema, obj: Any) -> core_schema.CoreSchema:
-    if obj == deque:
-        # bare `deque` type used as annotation
-        return _deque_any_schema()
+        if self.mapped_origin in (list, set, frozenset):
+            if self.mapped_origin is list:
+                constrained_schema = core_schema.list_schema(items_schema, **metadata)
+            elif self.mapped_origin is set:
+                constrained_schema = core_schema.set_schema(items_schema, **metadata)
+            else:
+                assert self.mapped_origin is frozenset  # safety check in case we forget to add a case
```

Comment:
might be more typesafe to do an `else: assert_never(self.mapped_origin)`, if that works anyway. Not a big deal

Reply:
I don’t think this part of the codebase is well typed enough for that to work